File: mobile_app/lib/models/overtime_model.dart
class OvertimeItem {
  final String id;
  final String staffId;
  final String fromDate;
  final String toDate;
  final String startTime;
  final String endTime;
  final double amountDay;
  final String reason;
  final String status; // Pending, Approved, Rejected
  final String? comment;
  final String? managerName;
  final String? branch;
  final String requestedAt;
  final String? approvedAt;
  final String? createdBy;

  OvertimeItem({
    required this.id,
    required this.staffId,
    required this.fromDate,
    required this.toDate,
    required this.startTime,
    required this.endTime,
    required this.amountDay,
    required this.reason,
    required this.status,
    this.comment,
    this.managerName,
    this.branch,
    required this.requestedAt,
    this.approvedAt,
    this.createdBy,
  });

  factory OvertimeItem.fromJson(Map<String, dynamic> json) {
    String fDate = '';
    if (json['fromDate'] != null) {
      fDate = json['fromDate'].toString().split('T')[0];
    } else if (json['from_date'] != null) {
      fDate = json['from_date'].toString().split('T')[0];
    }

    String tDate = fDate;
    if (json['toDate'] != null) {
      tDate = json['toDate'].toString().split('T')[0];
    } else if (json['to_date'] != null) {
      tDate = json['to_date'].toString().split('T')[0];
    }

    final double daysVal = json['amountDay'] != null
        ? (double.tryParse(json['amountDay'].toString()) ?? 0.0)
        : (json['amount_day'] != null
            ? (double.tryParse(json['amount_day'].toString()) ?? 0.0)
            : 0.0);

    String mName = '';
    if (json['manager'] != null && json['manager'] is Map) {
      mName = json['manager']['nameEn']?.toString() ?? json['manager']['nameKh']?.toString() ?? '';
    }
    if (mName.isEmpty && json['managerName'] != null) {
      mName = json['managerName'].toString();
    }

    String bName = '';
    if (json['branchLocation'] != null && json['branchLocation'] is Map) {
      bName = json['branchLocation']['name']?.toString() ?? '';
    }
    if (bName.isEmpty && json['branch'] != null) {
      bName = json['branch'].toString();
    }

    return OvertimeItem(
      id: json['id']?.toString() ?? '',
      staffId: json['staffId']?.toString() ?? json['staff_id']?.toString() ?? '',
      fromDate: fDate,
      toDate: tDate,
      startTime: json['startTime']?.toString() ?? json['start_time']?.toString() ?? '',
      endTime: json['endTime']?.toString() ?? json['end_time']?.toString() ?? '',
      amountDay: daysVal,
      reason: json['reason']?.toString() ?? '',
      status: json['status']?.toString() ?? 'Pending',
      comment: json['comment']?.toString(),
      managerName: mName.isNotEmpty ? mName : null,
      branch: bName.isNotEmpty ? bName : null,
      requestedAt: json['requestedAt']?.toString().split('T')[0] ?? json['created_at']?.toString().split('T')[0] ?? fDate,
      approvedAt: json['approvedAt']?.toString().split('T')[0] ?? json['approved_at']?.toString().split('T')[0],
      createdBy: json['createdBy']?.toString() ?? json['created_by']?.toString(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'staffId': staffId,
      'fromDate': fromDate,
      'toDate': toDate,
      'startTime': startTime,
      'endTime': endTime,
      'amountDay': amountDay,
      'reason': reason,
      'status': status,
      'comment': comment,
      'managerName': managerName,
      'branch': branch,
      'requestedAt': requestedAt,
      'approvedAt': approvedAt,
      'createdBy': createdBy,
    };
  }
}

class OvertimeStat {
  final int totalRequests;
  final int pendingRequests;
  final int approvedRequests;
  final int rejectedRequests;
  final double totalApprovedDays;

  OvertimeStat({
    required this.totalRequests,
    required this.pendingRequests,
    required this.approvedRequests,
    required this.rejectedRequests,
    required this.totalApprovedDays,
  });
}

File: frontend/src/models/OvertimeModel.js
import { BaseModel } from './BaseModel';
import { EmployeeModel } from './EmployeeModel';

/**
 * OvertimeModel
 * Encapsulates overtime request records, hours & day fraction calculations.
 */
export class OvertimeModel extends BaseModel {
  constructor(data = {}) {
    super(data);
    this.staffId = data.staffId || '';
    this.managerId = data.managerId || '';
    this.managerName = data.managerName || '';
    this.branch = data.branch || '';
    this.branchId = data.branchId || null;
    this.branchLocation = data.branchLocation || null;
    this.fromDate = data.fromDate ? new Date(data.fromDate) : null;
    this.toDate = data.toDate ? new Date(data.toDate) : null;
    this.startTime = data.startTime || '';
    this.endTime = data.endTime || '';
    this.amountDay = data.amountDay ? String(data.amountDay) : '0';
    this.reason = data.reason || '';
    this.status = data.status || 'Pending';
    this.comment = data.comment || '';
    this.requestedAt = data.requestedAt ? new Date(data.requestedAt) : null;
    this.approvedAt = data.approvedAt ? new Date(data.approvedAt) : null;
    this.createdBy = data.createdBy || '';
    this.employee = data.employee ? new EmployeeModel(data.employee) : null;
    this.manager = data.manager ? new EmployeeModel(data.manager) : null;
  }

  /**
   * Check if overtime is pending
   * @returns {boolean}
   */
  isPending() {
    return this.status === 'Pending';
  }

  /**
   * Check if overtime is approved
   * @returns {boolean}
   */
  isApproved() {
    return this.status === 'Approved';
  }

  /**
   * Status badge styling class
   * @returns {string}
   */
  getStatusBadgeClass() {
    switch (this.status) {
      case 'Approved':
        return 'bg-emerald-500/10 text-emerald-300 ring-emerald-500/20';
      case 'Rejected':
        return 'bg-rose-500/10 text-rose-300 ring-rose-500/20';
      default:
        return 'bg-amber-500/10 text-amber-300 ring-amber-500/20';
    }
  }

  /**
   * Factory method to create an array of models from API response
   * @param {Array} list
   * @returns {OvertimeModel[]}
   */
  static fromArray(list = []) {
    return Array.isArray(list) ? list.map(item => new OvertimeModel(item)) : [];
  }
}
